SO-4786: Add support for relationship concrete values

apeteri opened the pull request.

apeteri committed:
feat(taxonomy): Add key set support for internal ID maps and multimaps

apeteri committed:
feat(taxonomy): Update taxonomy building process

- Collect all stated relationships
- Collect non-IS A statement fragments extracted from OWL axioms
- Collect all stated axioms

apeteri committed:
feat(taxonomy): Update OWL ontology building process

- Parse and include OWL axioms as-is
- Generate an extra axiom on the fly for concepts that still have stated
relationships

apeteri committed:
chore(releng): Update SNOMED OWL Toolkit to 3.0.2

apeteri committed:
feat(domain): Add relationship value field to SnomedRelationship

apeteri committed:
feat(request): Support relationship value in create request and builder

apeteri committed:
feat(core): Pass relationship value to the document builder

apeteri committed:
fix(core): Rename moduleId "wither" method in SnomedComponentBuilder

apeteri committed:
feat(request): Support relationship value in update request and builder

apeteri committed:
fix(request): Rename moduleId updater method in...

...SnomedComponentUpdateRequest

apeteri committed:
feat(domain): Add equals/hashCode/toString impl. to RelationshipValue

apeteri committed:
feat(index): Add value fields to SnomedRelationshipIndexEntry

apeteri committed:
feat(request): Extend search support for relationship values

apeteri committed:
feat(snomed): Add RF2 header for relationship with value

apeteri committed:
feat(snomed): Relationships with values should not contribute null...

...destinationIds to module dependencies when versioning SNOMED CT

apeteri committed:
feat(snomed): Add RF2 importer support for relationships with value

apeteri committed:
feat(snomed): Populate value in SnomedRelationshipConverter; ignore...

...relationships that do not have a destination ID when expanding
destination concepts

apeteri committed:
feat(snomed): Add RF2 export support for relationships with value

apeteri committed:
feat(snomed): Update REST service for SNOMED CT relationships

Adds create, update and search support for relationships with a concrete
value.

apeteri committed:
feat(snomed): Add e2e test cases for concrete values

apeteri committed:
fix(snomed): Update error message assertion in SnomedConceptApiTest

apeteri committed:
fix(snomed): Don't create empty SnomedConcept instances when using...

...the SCTID setters on SnomedRelationship

apeteri committed:
fix(snomed): Use SnomedRelationshipIndexEntry field constant for sorting

apeteri committed:
fix(test): Fix test failures because of new RF2 file name

"sct2_Relationship" is also a prefix of
"sct2_RelationshipConcreteValue", which resulted in assertion failures.

apeteri committed:
fix(snomed): Only relationships with a destination ID should appear...

...in sct2_Relationship_ and sct2_StatedRelationship_ files

apeteri committed:
feat(snomed): Extend e2e test case with concrete value export

apeteri committed:
feat(snomed): Split StatementFragment into two subclasses

One for statements with a destination SCTID, one for values.

apeteri committed:
fix(snomed): Add separate field constants for relationship values

Field selection uses indexed field names at the moment, logical property
names are not supported.

apeteri committed:
fix(snomed): Remove boolean value type

This type is not supported in the OWL 2 EL profile; only types with a
(countably) infinite value space is supported, and allowed types can
either have 0 or infinite overlap.

apeteri committed:
feat(snomed): Allow creating OWL relationship document with value

apeteri committed:
feat(snomed): Use updated OWL toolkit to convert to-from...

...relationships with value

apeteri committed:
fix(index): Don't create "null" strings when mapping results to String[]

apeteri committed:
fix(snomed): Add missing destinationNegated flag to OWL relationship doc

apeteri committed:
feat(snomed): Add mapping function to StatementFragment and subclasses

apeteri committed:
feat(snomed): Update ReasonerTaxonomy and building process

- Use InternalSctIdSet instead of InternalIdMap for defining concepts
- Use InternalIdMultimap for storing ConcreteDomainFragments
- Replace Lists.newArrayListWithExpectedSize with new ArrayList(int)
- Deprecate "has stated pair" detection, as it produced false negatives

apeteri committed:
feat(snomed): Support relationships with value in NormalFormGenerator

- Allow creation of NormalFormValues from StatementFragmentWithValues
- Cast and map StatementFragment subclasses as appropriate
- Deprecate ConcreteDomainFragment generation - existing inferred CD
members will be inactivated, and inferred relationships with value will
be created in their place

apeteri committed:
feat(snomed): Restore OWL data properties, DataHasValue generation in...

...DelegateOntology

apeteri committed:
feat(snomed): Support relationships with value in classification...

...changes indexing and retrieval

apeteri committed:
feat(snomed): Create inferred relationships with value in SaveJobRequest

apeteri committed:
feat(snomed): Add test cases to SnomedClassificationApiTest

apeteri committed:
feat(index): Support indexing fields with type "double"

apeteri committed:
feat(snomed): Add expression evaluator method to RelationshipValue

Allows matching instances against other values with an operator
argument.

apeteri committed:
feat(snomed): Add method to RelationshipValue that returns a typed...

...Object instance, representing the value without any prefixes or
escaping.

apeteri committed:
feat(snomed): Index values stored on SnomedOWLRelationshipDocuments

This makes them searchable eg. for ECL evaluation purposes.

apeteri committed:
feat(snomed): Add support for relationships with value in ECL...

...refinement evaluation, both in OWL axiom and standalone form

apeteri committed:
fix(snomed): Small fixes around relationship-with-value handling

- Add utility method for determining if a relationship has a value
- Fix conversion to/from concrete values via the OWL Toolkit
- Add factory method for creating an OWL relationship from a regular
relationship
- Initialize union group to 0 in Rf2ConcreteValueContentType
- Initialize relationship value in Rf2TransactionContext

apeteri committed:
fix(snomed): Review uses of getDestinationId() on relationships

apeteri committed:
fix(snomed): Small fixes around relationship objects

- Decrease visibility of getters on SnomedOWLRelationshipDocument
- Move factory method to RelationshipValue
- Use getValueAsObject() method name consistently (getValue(), where
exists, will refer to the String literal form)

apeteri committed:
fix(snomed): Fix the reported conflicting concept in...

...SnomedComponentReferencingDetachedConceptRule; it should not be fixed
to a single concept ID (description type or relationship destination) if
the problem occurred on another concept (eg. description's case
significance).

apeteri committed:
feat(snomed): Add query expression for filtering OWL relationships...

...by destination ID

Its only caller is rule_mrcm_constraint, which should be reviewed, as
some of its constraints should be placed under a single nested query
clause instead of filtering independently.

apeteri committed:
fix(snomed): Rearrange order of parameters in factory method

apeteri committed:
fix(snomed): Add JsonProperty annotation to protected fields of...

...SnomedOWLRelationshipDocument

apeteri committed:
chore(releng): Update launch configuration of validation-snomed-tests

apeteri committed:
Merge remote-tracking branch 'origin/7.x' into feature/SO-4786_relationship_with_value

apeteri committed:
fix(snomed): Fix param tag in javadoc of SnomedRelationshipBuilder

apeteri committed:
fix(rf2): Remove boolean validation from Rf2ConcreteValueContentType

apeteri committed:
fix(rf2): Remove boolean validation message from Rf2ValidationDefects

apeteri committed:
fix(index): Use "hasDestinationId" as Expression name consistently

apeteri committed:
fix(snomed): Move "synthetic" VALUE_TYPE field to the non-RF2 section...

...in SnomedRelationship's "all fields" set

apeteri committed:
fix(snomed): Collect all non-IS A fragments from OWL axioms for the...

...normal form generation step, not all grouped fragments

apeteri committed:
fix(snomed): Fix OWL expression query in rule_mrcm_constraint

apeteri committed:
fix(snomed): Add hasDestinationId() to negative match in rule_stated_...

...mrcm_constraint; "does not have the following destination IDs"
clauses also match when the relationship has a concrete value otherwise.

cmark commented on snomed/com.b2international.snowowl.snomed.core.rest/src/com/b2international/snowowl/snomed/core/rest/domain/SnomedRelationshipRestSearch.java:
```diff
@@ -44,6 +46,15 @@
+	private SearchResourceRequest.Operator operator;
+
+	@ApiParam(value = "The value to match (in literal form)")
+	private String value;
```
Multi-value support?

apeteri replied:
I'd keep it single valued for now. Multi-values are not great on the request level either.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/ecl/SnomedEclRefinementEvaluator.java:
```diff
@@ -487,6 +495,169 @@ public SnomedEclRefinementEvaluator(EclExpression focusConcepts) {
+
+		// TODO: does this request need to support filtering by group?
+		Promise<Collection<Property>> relationships = SnomedRequests.prepareSearchRelationship()
+				.all()
```
To reduce memory-pressure and improve speed, use field selection here to reduce the number of fields loaded from the concrete value relationship. Same as loading fields during standard relationship search.

apeteri committed:
fix(snomed): Add missing build() call to ExpressionBuilders

apeteri commented:
Ready for review; the commits added today are the results of a self-review I have personally conducted. 😄

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedRelationshipSearchRequestBuilder.java:
```diff
@@ -86,9 +88,36 @@ public SnomedRelationshipSearchRequestBuilder filterByModifier(Iterable<String>
+		return filterByValue(SearchResourceRequest.Operator.EQUALS, value);
+	}
+	
+	public SnomedRelationshipSearchRequestBuilder filterByValue(Iterable<RelationshipValue> values) {
```
It is still a bit unsettled, but we have started using pluralized method names (`filterByValues`) for multi-valued filters to keep them separated from the singular versions.

cmark commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedRelationshipUpdateRequest.java:
```diff
@@ -15,80 +15,98 @@
+	// Not @NotNull, update of relationship value is optional
+	private RelationshipValue value;
+
+	@Min(0)	@Max(Integer.MAX_VALUE)
```
A bit of inconsistency between the create and update requests. Create allows only `Byte.MAX_VALUE` while update allows `Integer.MAX_VALUE` `groups` and `unionGroups`.

apeteri committed:
fix(snomed): Use field selection in relationship search request of...

...SnomedEclRefinementEvaluator

apeteri committed:
fix(snomed): Change filter method names that accept multiple values...

...in SnomedRelationshipSearchRequestBuilder to plural form

apeteri committed:
fix(snomed): Limit maximum group number to 32767 in relationship...

...create and update requests. The type remains `int` as it is easier to
work with in both persistence and runtime scenarios.

apeteri committed:
fix(tests): Add test cases for all relationship value types to...

...SnomedClassificationApiTest

apeteri committed:
fix(snomed): Allow all non-negative integers for relationship groups

apeteri commented:
I've made the requested changes.

apeteri committed:
chore: Add new feature to README.md

apeteri committed:
fix(reasoner): Add "hasValue" property to ReasonerRelationship

apeteri committed:
fix(reasoner): Add both "Sub<Object|Data>PropertyOf" and "SubClassOf"...

...axiom for object and data properties, to support taxonomy extraction

cmark reviewed (approved):
LGTM!
